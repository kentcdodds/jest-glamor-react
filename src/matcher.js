const chalk = require('chalk')
const diff = require('jest-diff')
const stripAnsi = require('strip-ansi')
const {toMatchSnapshot} = require('jest-snapshot')

const isAddition = line => /^\+/.test(line)

const isDeletion = line => /^-/.test(line)

const colorize = message => {
  const messageLines = message.split('\n')

  return messageLines
    .map(line => {
      if (isAddition(line)) {
        return chalk.red(line)
      }
      if (isDeletion(line)) {
        return chalk.green(line)
      }
      return chalk.dim(line)
    })
    .join('\n')
}

const matcher = {
  toMatchSnapshotWithGlamor(...args) {
    const result = toMatchSnapshot.apply(this, args)
    let message

    if (!result.pass) {
      message = diff(result.expected, result.actual, {
        aAnnotation: 'Snapshot',
        bAnnotation: 'Received',
      })
      message = stripAnsi(message)
      message = colorize(message)
    }

    return {pass: result.pass, message}
  },
}

module.exports = matcher
