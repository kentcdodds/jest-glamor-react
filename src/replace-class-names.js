const replaceClassNames = (selectors, styles, code) => {
  let index = 0
  return selectors.reduce((acc, className) => {
    if (className.indexOf('.css-') === 0) {
      const escapedRegex = new RegExp(
        className.replace('.', '').replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&'),
        'g',
      )
      return acc.replace(escapedRegex, `glamor-${index++}`)
    }
    return acc
  }, `${styles}\n\n${code}`)
}

module.exports = {replaceClassNames}
