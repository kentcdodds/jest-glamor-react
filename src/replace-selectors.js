const replaceSelectors = (selectors, styles, code) => {
  // let index = 0
  const selectorIds = selectors.reduce((acc, selector) => {
    if (selector.includes('[data-css-')) {
      acc.add(/\[data-css-(.*)\]/g.exec(selector)[1])
    } else if (selector.includes('.css-')) {
      acc.add(/\.css-([^ >]*)/g.exec(selector)[1])
    }
    return acc
  }, new Set())
  return Array.from(selectorIds).reduce((string, id, index) => {
    return string.split(id).join(index)
  }, `${styles}\n\n${code}`)
}

module.exports = {replaceSelectors}
