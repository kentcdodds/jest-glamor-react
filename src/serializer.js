const {replaceClassNames} = require('./replace-class-names')
const {getNodes, markNodes, getSelectors, getStyles} = require('./utils')

function createSerializer(styleSheet, classNameReplacer) {
  function test(val) {
    return val &&
      !val.withStyles &&
      val.$$typeof === Symbol.for('react.test.json')
  }

  function print(val, printer) {
    const nodes = getNodes(val)
    markNodes(nodes)
    const selectors = getSelectors(nodes)
    const styles = getStyles(selectors, styleSheet)
    const printedVal = printer(val)
    if (styles) {
      return replaceClassNames(
        selectors,
        styles,
        printedVal,
        classNameReplacer,
      )
    } else {
      return printedVal
    }
  }

  return {test, print}
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getGlamorStyleSheet = () => require('glamor').styleSheet
const glamorSerializer = createSerializer(getGlamorStyleSheet)
createSerializer.test = glamorSerializer.test
createSerializer.print = glamorSerializer.print

module.exports = createSerializer
