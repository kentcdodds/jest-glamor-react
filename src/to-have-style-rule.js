const {getAST, getClassNames} = require('./utils')

const getGlamorStyleSheet = () => require('glamor').styleSheet

function valueMatches(declaration, value) {
  return value instanceof RegExp ?
    value.test(declaration.value) :
    value === declaration.value
}

function toHaveStyleRule(received, property, value) {
  const selectors = getClassNames(received)
  const ast = getAST(selectors, getGlamorStyleSheet())

  const classNames = getClassNames(received)
  const rules = ast.stylesheet.rules.filter(rule =>
    classNames.some(cn => rule.selectors.includes(`.${cn}`)))

  const declaration = rules
    .reduce((decs, rule) => Object.assign([], decs, rule.declarations), [])
    .filter(dec => dec.type === 'declaration' && dec.property === property)
    .pop()

  if (!declaration) {
    return {
      pass: false,
      message: () => `Property not found: ${property}`,
    }
  }

  const pass = valueMatches(declaration, value)

  const message = () =>
    `Expected ${property}${pass ? ' not ' : ' '}to match:\n` +
    `  ${value}\n` +
    'Received:\n' +
    `  ${declaration.value}`

  return {
    pass,
    message,
  }
}

module.exports = {
  toHaveStyleRule,
}
