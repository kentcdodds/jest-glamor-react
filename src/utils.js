const css = require('css')

function getNodes(node, nodes = []) {
  if (node.children) {
    const children =
      typeof node.children === 'function' ? node.children() : node.children
    if (Array.isArray(children)) {
      children.forEach(child => getNodes(child, nodes))
    }
  }

  if (typeof node === 'object') {
    nodes.push(node)
  }

  return nodes
}

function markNodes(nodes) {
  nodes.forEach(node => {
    node.serializedWithJestGlamorReact = true
  })
}

function getSelectors(nodes) {
  return nodes.reduce((selectors, node) => {
    if (node instanceof global.HTMLElement) {
      return getSelectorsFromDOM(selectors, node)
    } else {
      return getSelectorsFromProps(selectors, node)
    }
  }, [])
}

// eslint-disable-next-line
function getSelectorsFromProps(selectors, node) {
  const props = typeof node.props === 'function' ? node.props() : node.props
  if (!props) return []
  const className = props.className || props.class
  if (className) {
    selectors = selectors.concat(
      className
        .toString()
        .split(' ')
        .map(cn => `.${cn}`),
    )
  }
  const dataProps = Object.keys(props).reduce((dProps, key) => {
    if (key.startsWith('data-')) {
      dProps.push(`[${key}]`)
    }
    return dProps
  }, [])
  if (dataProps.length) {
    selectors = selectors.concat(dataProps)
  }
  return selectors
}

function getSelectorsFromDOM(selectors, node) {
  const root = document.createElement('div')
  root.appendChild(node)
  const allChildren = root.querySelectorAll('*')
  selectors = Array.from(allChildren).reduce((s, c) => {
    s.push(...Array.from(c.classList).map(cn => `.${cn}`))
    s.push(
      ...c
        .getAttributeNames()
        .filter(key => key.startsWith('data-'))
        .map(key => `[${key}]`),
    )
    return s
  }, selectors)
  return selectors
}

function filterChildSelector(baseSelector) {
  if (baseSelector.slice(-1) === '>') {
    return baseSelector.slice(0, -1)
  }
  return baseSelector
}

function getAST(nodeSelectors, styleSheet) {
  const tags =
    typeof styleSheet === 'function' ? styleSheet().tags : styleSheet.tags
  const styles = tags
    .map(tag => /* istanbul ignore next */ tag.textContent || '')
    .join('\n')
  const ast = css.parse(styles)
  const rules = ast.stylesheet.rules.filter(filter)
  const mediaQueries = getMediaQueries(ast, filter)

  ast.stylesheet.rules = [...rules, ...mediaQueries]

  return ast

  function filter(rule) {
    if (rule.type === 'rule') {
      return rule.selectors.some(selector => {
        const baseSelector = filterChildSelector(
          selector.split(/:| |\./).filter(s => !!s)[0],
        )
        return nodeSelectors.some(
          sel => sel === baseSelector || sel === `.${baseSelector}`,
        )
      })
    }
    return false
  }
}

function getStylesAndAllSelectors(nodeSelectors, styleSheet) {
  const ast = getAST(nodeSelectors, styleSheet)
  const allSelectors = ast.stylesheet.rules.reduce((s, r) => {
    if (r.selectors) {
      s.push(...r.selectors)
    }
    return s
  }, [])
  const styles = css.stringify(ast)
  return {styles, allSelectors}
}

function getMediaQueries(ast, filter) {
  return ast.stylesheet.rules
    .filter(rule => rule.type === 'media' || rule.type === 'supports')
    .reduce((acc, mediaQuery) => {
      mediaQuery.rules = mediaQuery.rules.filter(filter)

      if (mediaQuery.rules.length) {
        return acc.concat(mediaQuery)
      }

      return acc
    }, [])
}

function shouldDive(node) {
  return typeof node.dive === 'function' && typeof node.type() !== 'string'
}

function isTagWithClassName(node) {
  return node.prop('className') && typeof node.type() === 'string'
}

function getClassNameFromEnzyme(received) {
  const tree = shouldDive(received) ? received.dive() : received
  const components = tree.findWhere(isTagWithClassName)
  return components.length && components.first().prop('className')
}

function getClassNameFromTestRenderer(received) {
  return received.props.className || received.props.class
}

function getClassNameFromCheerio(received) {
  /* istanbul ignore next */
  if (received[0].type === 'root') {
    // I'm honestly not sure we need this, but it was here
    // and I don't want to break it...
    return received
      .children()
      .first()
      .attr('class')
  } else {
    return received.attr('class')
  }
}

// eslint-disable-next-line complexity
function getClassNames(received) {
  let className

  if (received) {
    if (received instanceof HTMLElement) {
      className = received.className
    } else if (received.$$typeof === Symbol.for('react.test.json')) {
      className = getClassNameFromTestRenderer(received)
    } else if (typeof received.findWhere === 'function') {
      className = getClassNameFromEnzyme(received)
    } else if (received.cheerio === '[cheerio object]') {
      className = getClassNameFromCheerio(received)
    }
  }

  return className ? className.split(/\s/) : []
}

module.exports = {
  getNodes,
  markNodes,
  getSelectors,
  getAST,
  getStylesAndAllSelectors,
  getClassNames,
}
