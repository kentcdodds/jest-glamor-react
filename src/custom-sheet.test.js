import React from 'react'
import renderer from 'react-test-renderer'
import cxs, {sheet} from 'cxs'
import * as enzyme from 'enzyme'
import toJson from 'enzyme-to-json'
import {matcher, serializer} from '../src'

expect.addSnapshotSerializer(serializer(sheet))
expect.extend(matcher)

function Wrapper(props) {
  const className = cxs({
    padding: '4em',
    background: 'papayawhip',
  })
  return <section className={`${className}`} {...props} />
}

function Title(props) {
  const className = cxs({
    fontSize: '1.5em',
    textAlign: 'center',
    color: 'palevioletred',
  })
  return <h1 className={`${className}`} {...props} />
}

test('react-test-renderer', () => {
  const tree = renderer
    .create(
      <Wrapper>
        <Title>Hello World, this is my first glamor styled component!</Title>
      </Wrapper>,
    )
    .toJSON()

  expect(tree).toMatchSnapshotWithGlamor()
})

test('enzyme', () => {
  const ui = (
    <Wrapper>
      <Title>Hello World, this is my first glamor styled component!</Title>
    </Wrapper>
  )

  const enzymeMethods = ['shallow', 'mount', 'render']
  enzymeMethods.forEach(method => {
    const tree = enzyme[method](ui)
    expect(toJson(tree)).toMatchSnapshotWithGlamor(`enzyme.${method}`)
  })
})

test('works when the root element does not have styles', () => {
  const tree = renderer
    .create(
      <div>
        <Wrapper />
      </div>,
    )
    .toJSON()

  expect(tree).toMatchSnapshotWithGlamor()
})

test(`doesn't mess up stuff that does't have styles`, () => {
  const tree = renderer.create(<div />).toJSON()

  expect(tree).toMatchSnapshotWithGlamor()
})

const generalTests = [
  {
    title: 'media queries',
    styles: {
      fontSize: 24,
      margin: 12,
      '@media (max-width: 641px)': {
        fontSize: 20,
        margin: 10,
      },
    },
  },
  {
    title: 'pseudo elements',
    styles: {
      '& button': {
        color: 'green',
      },
    },
  },
  {
    title: 'pseudo states',
    styles: {
      ':hover': {
        backgroundColor: 'blue',
      },
    },
  },
]

generalTests.forEach(({title, styles}, index) => {
  test(title, () => {
    const tree = renderer.create(<div className={cxs(styles)} />)
    expect(tree).toMatchSnapshotWithGlamor(
      `${index + 1}. general tests: ${title} - ${JSON.stringify(styles)}`,
    )
  })
})
