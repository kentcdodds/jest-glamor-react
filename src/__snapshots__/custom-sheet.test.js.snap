// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. general tests: media queries - {"fontSize":24,"margin":12,"@media (max-width: 641px)":{"fontSize":20,"margin":10}} 1`] = `
.fs-24 {
  font-size: 24px;
}

.m-12 {
  margin: 12px;
}

@media (max-width: 641px) {
  ._9gzisy {
    font-size: 20px;
  }
}

@media (max-width: 641px) {
  ._d2w6i1 {
    margin: 10px;
  }
}

<div
  className="fs-24 m-12 _9gzisy _d2w6i1"
/>
`;

exports[`2. general tests: pseudo elements - {"& button":{"color":"green"}} 1`] = `
.___button-c-green & button {
  color: green;
}

<div
  className="___button-c-green"
/>
`;

exports[`3. general tests: pseudo states - {":hover":{"backgroundColor":"blue"}} 1`] = `
._hover-bc-blue:hover {
  background-color: blue;
}

<div
  className="_hover-bc-blue"
/>
`;

exports[`doesn't mess up stuff that does't have styles 1`] = `<div />`;

exports[`enzyme.mount 1`] = `
.p-4em {
  padding: 4em;
}

.b-papayawhip {
  background: papayawhip;
}

.fs-1p5em {
  font-size: 1.5em;
}

.text-align-center {
  text-align: center;
}

.c-palevioletred {
  color: palevioletred;
}

<Wrapper>
  <section
    className="p-4em b-papayawhip"
  >
    <Title>
      <h1
        className="fs-1p5em text-align-center c-palevioletred"
      >
        Hello World, this is my first glamor styled component!
      </h1>
    </Title>
  </section>
</Wrapper>
`;

exports[`enzyme.render 1`] = `
.p-4em {
  padding: 4em;
}

.b-papayawhip {
  background: papayawhip;
}

.fs-1p5em {
  font-size: 1.5em;
}

.text-align-center {
  text-align: center;
}

.c-palevioletred {
  color: palevioletred;
}

<section
  class="p-4em b-papayawhip"
>
  <h1
    class="fs-1p5em text-align-center c-palevioletred"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;

exports[`enzyme.shallow 1`] = `
.p-4em {
  padding: 4em;
}

.b-papayawhip {
  background: papayawhip;
}

<section
  className="p-4em b-papayawhip"
>
  <Title>
    Hello World, this is my first glamor styled component!
  </Title>
</section>
`;

exports[`react-test-renderer 1`] = `
.p-4em {
  padding: 4em;
}

.b-papayawhip {
  background: papayawhip;
}

.fs-1p5em {
  font-size: 1.5em;
}

.text-align-center {
  text-align: center;
}

.c-palevioletred {
  color: palevioletred;
}

<section
  className="p-4em b-papayawhip"
>
  <h1
    className="fs-1p5em text-align-center c-palevioletred"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;

exports[`works when the root element does not have styles 1`] = `
.p-4em {
  padding: 4em;
}

.b-papayawhip {
  background: papayawhip;
}

<div>
  <section
    className="p-4em b-papayawhip"
  />
</div>
`;
